### DISCUSSION TITLE
Keep type information when wrapping prisma call in a function

### DISCUSSION THREAD

**Discussion by redsuperbat** [prisma/prisma#25102] [Q&A]
### Question

Is there a way to keep the type inference of the prisma types when wrapping a prisma call in a function/method.

For example, here I'm defining a method signature which wraps a prisma call and merges a where clause to the query. 
That makes sure that all of my order rows are active and that my order has not expired:

```ts
  getOrderRows<T extends InternalArgs>(
    args: Prisma.OrderRowFindManyArgs<T> = {},
  ) {
    const constantWhere: Prisma.OrderRowWhereInput = {
      state: "ACTIVE",
      order: {
        effectiveEndDate: {
          gte: new Date(),
        },
      },
    };
    args.where = merge(constantWhere, args.where);
    return this.prisma.orderRow.findMany(args);
  }
```

I want to be able to pass any other prisma argument to the method and still get the type inference from includes/etc.

Is it possible? With this code above it does not work. The type inference is not kept.

Thanks for your help.

**Comment by galenwinsor** [comment]
Try this:
```typescript
const constantWhere = {
      state: "ACTIVE",
      order: {
        effectiveEndDate: {
          gte: new Date(),
        },
      },
    } satisfies Prisma.OrderRowWhereInput;
```
so that the type isn't broadened.

Where does `merge` come from? Depending on that function's implementation, it may or may not preserve the type of `args.where`.

**Comment by ludralph** [comment]
Hi @redsuperbat 👋 

You can also use Prisma's client extensions to create custom methods while preserving type inference. This approach allows you to define custom queries and still benefit from Prisma's type safety.
```js
import { PrismaClient, Prisma } from '@prisma/client';
import merge from 'lodash.merge';

const prisma = new PrismaClient();

const prismaB = prisma.$extends({
  model: {
    orderRow: {
      async getOrderRows<T extends Prisma.OrderRowFindManyArgs>(
        args: T
      ): Promise<Prisma.Result<Prisma.OrderRowFindManyArgs, T, 'findMany'>> {
        const constantWhere: Prisma.OrderRowWhereInput = {
          state: 'ACTIVE',
          order: {
            effectiveEndDate: {
              gte: new Date(),
            },
          },
        };
        args.where = merge(constantWhere, args.where);
        return await prisma.orderRow.findMany(args);
      },
    },
  },
});

// Usage
async function main() {
  const orderRows = await prismaB.orderRow.getOrderRows({
    where: { someOtherCondition: true },
    include: { order: true },
  });

  console.log(orderRows);
}

main().catch((e) => {
  console.error(e);
  prisma.$disconnect();
});
```
We use `prisma.$extends` to add a custom method getOrderRows to the orderRow model. The method `getOrderRows` merges a constant `where` clause with the provided arguments and calls `prisma.orderRow.findMany`. Type inference is preserved for the arguments and the return type. With this, you can pass any Prisma arguments to the method and still get the type inference from includes, selects, etc.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
